File: 300.最长递增子序列.py
#
# @lc app=leetcode.cn id=300 lang=python
#
# [300] 最长递增子序列
#
# 解法1(T46% S30%): 一点点动态规划的思想但不完全是，当前一个字符能不能是最长增序列的一员取决于他之前有没有比他小的，到它的长度=之前比他小中最大的长度+1；最终返回的就是整个dp数组中最大的长度，复杂度是O(N^2)

# @lc code=start
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp = [1 for _ in range(len(nums))]
        dp[0] = 1

        for i in range(1, len(nums)):
            for j in range(i-1, -1, -1):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j]+1)

        return max(dp)
# @lc code=end

File: javascript/300.最长递增子序列.js
/*
 * @lc app=leetcode.cn id=300 lang=javascript
 *
 * [300] 最长递增子序列
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    var dp = new Array(nums.length).fill(1);
    for (var i=1; i<nums.length; i++) {
        var j=i-1;
        while (j>-1) {
            if (nums[j] < nums[i]) {
                dp[i] = Math.max(dp[i], dp[j]+1);
            }
            j--;
        }
    }
    return Math.max(...dp);
};
// @lc code=end

